x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java

```diff
@@ -141,6 +154,11 @@ public String getIndex() {
             return index;
         }
 
+        public long estimateMemoryUsageBytes() {
+            // While loading the model in the process we need twice the model size
+            return MEMORY_OVERHEAD.getBytes() + 2 * modelBytes;
```

Comment:
While I think this is a good estimate, I am not sure its going to work well with autoscaling. 

If you think about it, the memory a node needs is actually `MEMORY_OVERHEAD.getBytes() * numModels + largestModelSize + largestModelSize + model_a +...+model_n`

When we load the models, we should do that serially, and always require the scaling to have at least the overhead required to load each model. Always requiring 2x EVERY model is way too much memory.

Reply:
We decided to use 2x EVERY model to avoid the risk of flapping with autoscaling, or mind-boggling complexity in what we report as our memory requirement to autoscaling.

At smaller node sizes autoscaling the ML tier will replace every ML node with a bigger one.  But that bigger one would have to have enough space to reload the models.

The long-term plan for avoiding the inefficiency is to send the model from Java to C++ on disk, so it doesn't seem worth adding all the extra complexity now only to take it away when we switch to disk.

Reply:
> We decided to use 2x EVERY model to avoid the risk of flapping with autoscaling, or mind-boggling complexity in what we report as our memory requirement to autoscaling.

If loaded serially, we don't need this. But I suppose we can do that for now. We won't flap with autoscaling if the node memory estimate simply adds the "larges model size" to whatever the true model size is + overhead.

Reply:
If we ensuring one model is loaded at a time then we can indeed calculate the total node memory load as `sum(started_model_sizes) + max(started_model_sizes)`. This isn't too complex. We basically make each task report the model size as it's memory usage and we change the `NodeLoadDetector` to add the max model size into the total. I'm happy to go for this if we think this is adequate and that we don't then implement the disk approach at all. If we still think we'll eventually load the model via disk, then there's no much point changing it.

Reply:
Offline discussion led to the conclusion that we will indeed move to a different way of doing this in the future so for now the 2x estimate is good enough.

Reply:
I would like to keep the `sum(started_model_sizes) + max(started_model_sizes)` option alive. The extra complexity is not too onerous and pays off with a much better estimate of memory usage. The alternatives themselves are very complex and we don't have a timeframe when they might be completed.

Reply:
> The extra complexity is not too onerous

We have a parallel project running to change the way model allocation is done so that the models run on all nodes of a certain type in the cluster instead of just one.

The way this works with the current single persistent task and the way it works with the new multi-allocation is quite different.  I think it will be harder than expected to get this right for both cases.

We can always make it better in a followup PR.  I think it's better to get this PR merged as a step in the right direction in the knowledge we have to do more in this area in the future.

Reply:
We can easily change this later if we don't get to work on loading the model from disk.

Reply:
No I wasn't suggesting that the change goes in this PR only that the option may become more attractive in the medium term. However we move forward consistent reporting of memory usage- whether it be from an allocation or a single persistent task- will be key.